crucible-go/src/Lang/Crucible/Go/Types.hs

```diff
@@ -0,0 +1,326 @@
+addFunction nm cfg = modify $ \ts -> ts { functions = (nm, cfg) : functions ts }
+
+-- | Our generator user state only contains the local lexical
+-- environment. Gamma could be generalized to a stack of nested scopes
+-- to enable more efficient translation of assignment statements (not
+-- always introducing new variables when they already exist IN THE
+-- INNERMOST SCOPE).
+data GenState s =
+  GenState
+  { -- | lexical environment
+    gamma :: HashMap Text (SomeGoReg s)
+    -- | map from label text to crucible label
+  , label_map :: HashMap Text (Gen.Label s)
+    -- | stack of loop labels (continue, break)
+  , loop_labels :: [(Gen.Label s, Gen.Label s)]
+    -- | map from loop label to continue and break crucible labels
+  , loop_label_map :: HashMap Text (Gen.Label s, Gen.Label s)
+    -- | set when translating a labeled statement
+  , current_label :: Maybe Ident
+  }
+
+instance Default (GenState s) where
+  def = GenState { gamma = HM.empty
+                 , label_map = HM.empty
+                 , loop_labels = []
+                 , loop_label_map = HM.empty
+                 , current_label = Nothing }
+
+-- | The type of Go generator actions.
+type GoGenerator s ret a = Gen.Generator Go s GenState ret IO a
+data SomeGoGenerator s a =
+  forall ret. SomeGoGenerator (TypeRepr ret) (GoGenerator s ret a)
+instance Show (SomeGoGenerator s a) where
+  show = const "<SomeGoGenerator>"
+
+-- | A GoExpr is a value expression, possibly accompanied by a
+-- corresponding assignable location (where the value lives).
+data GoExpr s tp where
+  GoExpr :: Maybe (GoLoc s tp)
+         -> Gen.Expr Go s tp
+         -> GoExpr s tp
+  deriving Show
+instance C.ShowF (GoExpr s) where
+  showF = show
+type SomeGoExpr s = Some (GoExpr s)
+
+-- | Create a GoExpr with no location.
+mkGoExpr :: Gen.Expr Go s tp -> GoExpr s tp
+mkGoExpr = GoExpr Nothing
+
+mkSomeGoExpr :: Gen.Expr Go s tp -> SomeGoExpr s
+mkSomeGoExpr = Some . mkGoExpr
+
+mkSomeGoExpr' :: C.App Go (Gen.Expr Go s) tp -> SomeGoExpr s
+mkSomeGoExpr' = mkSomeGoExpr . Gen.App
+
+goExpr :: GoExpr s tp -> Gen.Expr Go s tp
+goExpr (GoExpr _loc e) = e
+
+goExprType :: GoExpr a tp -> TypeRepr tp
+goExprType (GoExpr _loc e) = exprType e
+
+-- | Pair of things with the same type index.
+data PairIx (f :: * -> CrucibleType -> *) s tp where
+  PairIx :: forall f s tp. f s tp -> f s tp -> PairIx f s tp
+
+-- | Pair of things with the same type index.
+type SomePairIx f s = Some (PairIx f s)
+type SomeGoPair s = Some (PairIx GoExpr s)
+type SomeExprPair s = Some (PairIx (Gen.Expr Go) s)
+
+
+-- | The translator monad.
+type TranslateM' a = StateT TransState (ExceptT String IO) a
+
+-- | A newtype around TranslateM' indexed by NodeTypes so it can be
+-- used as a functor in recursion constructions (see
+-- Lang.Crucible.Go.Rec).
+newtype TranslateM (tp :: NodeType) =
+  TranslateM { runTranslateM :: TranslateM' (Translated tp) }
+
+liftIO' :: IO a -> TranslateM' a
+liftIO' = lift . lift
```

Comment:
Why not just inline the definition of `TranslateM'` and then derive `MonadIO`?

Reply:
It seems that `TranslateM'` is used elsewhere, too. Maybe that could be cleaned up, but it would be a bunch of changes.